### PR TITLE
Implement wireguard for linux + fixes in executors/ssh client

### PR DESCRIPTION
**Pull request by grimpy**
Signed-off-by: Jo De Boeck <deboeck.jo@gmail.com>

### CODE REVIEW

File: JumpscaleCore/clients/ssh/SSHClientBase.py [hunk 1]
```diff
@@ -429,11 +425,11 @@ def _execute_script(
             content += "\n"
 
         if die:
-            content = "set -ex\n{}".format(content)
+            content = "set -e\n{}".format(content)
```
Comment (xmonader) [inline JumpscaleCore/clients/ssh/SSHClientBase.py]: not interested in the debug output?

File: JumpscaleCore/tools/executor/ExecutorBase.py [hunk 2]
```diff
@@ -119,47 +119,28 @@ def file_read(self, path):
...
-                cmd += "chmod %s %s\n" % (mode, path)
-            if owner:
-                cmd += "chown %s %s\n" % (owner, path)
-            if group:
-                cmd += "chgrp %s %s\n" % (group, path)
-
-            res = self.execute(cmd, showout=False, script=False, interactive=False)
+        temp = j.sal.fs.getTmpFilePath()
+        j.sal.fs.writeFile(filename=temp, contents=content, append=False)
+        self.upload(temp, path)
```
Comment (xmonader) [inline JumpscaleCore/tools/executor/ExecutorBase.py]: put file has mode parameter do we need to use chmod?

File: JumpscaleCore/clients/ssh/SSHClientBase.py [hunk 3]
```diff
@@ -429,11 +425,11 @@ def _execute_script(
...
 
         dpath = "/tmp/jsxssh_{}.sh".format(j.data.idgenerator.generateRandomInt(1, 100000))
 
-        self.executor.file_write(dpath, content, mode="7770")
+        self.executor.file_write(dpath, content, mode="777")
```
Comment (xmonader) [inline JumpscaleCore/clients/ssh/SSHClientBase.py]: to be honest i really believe mode should be int, to be consistent with ssh libs apis and what users expect

File: JumpscaleCore/tools/executor/ExecutorSSH.py [hunk 4]
```diff
@@ -21,14 +20,19 @@ def __init__(self, sshclient):
...
+        if env or sudo or script:
+            raise j.exceptions.NotFound("Not implemented for ssh execuor")
+        return self.sshclient.execute(cmd, interactive, showout, replace, die, timeout)
 
-    # def executeRaw(self, cmds, die=True, showout=True, timeout=120):
-    #     rc, out, err = self.sshclient.execute(cmds, die=die, showout=showout, timeout=timeout)
-    #     return rc, out, err
+    @property
+    def uid(self):
+        return self.sshclient._id
```
Comment (xmonader) [inline JumpscaleCore/tools/executor/ExecutorSSH.py]: maybe better to use uid property github.com/threefoldtech/jumpscaleX_core/blob/a423e95f94/JumpscaleCore/clients/ssh/SSHClientBase.py#L86

File: JumpscaleCore/tools/wireguard/WireGuard.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
+from Jumpscale import j
+from configparser import ConfigParser
+from io import StringIO
+
+def to_section(section, data):
```
Comment (xmonader) [inline JumpscaleCore/tools/wireguard/WireGuard.py]: maybe inifile tool would be better to use not sure

File: JumpscaleCore/tools/wireguard/WireGuard.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        config += to_section("Interface", interface)
+        for peerobject in self.peers_objects:
+            _, publickey = peerobject.key_pair_get()
+            peer = {"PublicKey": publickey,
+                    "AllowedIPs": peerobject.network_private}
+            if peerobject.network_public:
+                peer["EndPoint"] = f"{peerobject.network_public}:{peerobject.port}"
+            config += to_section("Peer", peer)
+
+        configpath = "/tmp/{}.conf".format(self.interface_name)
```
Comment (xmonader) [inline JumpscaleCore/tools/wireguard/WireGuard.py]: why not to stick to f strings

Comment (zaibon) [reply JumpscaleCore/tools/wireguard/WireGuard.py]: JSX requires python3.8 now ?

File: JumpscaleCore/clients/ssh/SSHClientBase.py (hunk 1 above)
Comment (grimpy) [reply JumpscaleCore/clients/ssh/SSHClientBase.py]: Nope as it ruines the stdout returned by  the command

File: JumpscaleCore/tools/executor/ExecutorSSH.py (hunk 4 above)
Comment (grimpy) [reply JumpscaleCore/tools/executor/ExecutorSSH.py]: ok

File: JumpscaleCore/tools/wireguard/WireGuard.py (hunk 6 above)
Comment (grimpy) [reply JumpscaleCore/tools/wireguard/WireGuard.py]: ok

**Comment by xmonader** [comment]
Thanks! looks great!

**Pull request merged**